esutil/bulk_indexer.go

```diff
@@ -0,0 +1,535 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package esutil
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/elastic/go-elasticsearch/v8"
+	"github.com/elastic/go-elasticsearch/v8/esapi"
+)
+
+// BulkIndexer represents a parallel, asynchronous, efficient indexer for Elasticsearch.
+//
+type BulkIndexer interface {
+	// Add adds an item to the indexer. It returns an error when the item cannot be added.
+	// Use the OnSuccess and OnFailure callbacks to get the operation result for the item.
+	//
+	// You must call the Close() method after you're done adding items.
+	//
+	// It is safe for concurrent use. When it's called from goroutines,
+	// they must finish before the call to Close, eg. using sync.WaitGroup.
+	Add(context.Context, BulkIndexerItem) error
+
+	// Close waits until all added items are flushed and closes the indexer.
+	Close(context.Context) error
+
+	// Stats returns indexer statistics.
+	Stats() BulkIndexerStats
+}
+
+// BulkIndexerConfig represents configuration of the indexer.
+//
+type BulkIndexerConfig struct {
+	NumWorkers    int           // The number of workers. Defaults to runtime.NumCPU().
+	FlushBytes    int           // The flush threshold in bytes. Defaults to 5MB.
+	FlushInterval time.Duration // The flush threshold as duration. Defaults to 10sec.
+
+	Client      *elasticsearch.Client   // The Elasticsearch client.
+	Decoder     BulkResponseJSONDecoder // A custom JSON decoder.
+	DebugLogger BulkIndexerDebugLogger  // An optional logger for debugging.
+
+	OnError func(error) // Called for indexer errors.
```

Comment:
Seeing as you have no context around what part of the indexer is emitting an error, what can you do with the error other than log it? This doesn't seem particularly useful to me - I'd probably just extend the logger to also log errors, or perhaps just log them as debug messages.

On the other hand, it might be helpful to know when a _bulk request has failed overall, or how long bulk requests are taking, etc. I think it might be useful to add another interface like follows to BulkIndexerConfig:

```go
type BulkIndexerFlusher interface {
    // Writer is used for writing the bulk operations.
    io.Writer

    // Flush sends a bulk request with the written bulk operations,
    // and then clears the buffer.
    Flush(context.Context) (*esapi.Response, error)
}

type DefaultBulkIndexerFlusher struct {
    Client *elasticsearch.Client
  
    buf bytes.Buffer

    // Parameters of the Bulk API.
    Index               string
    ErrorTrace          bool
    FilterPath          []string
    Header              http.Header
    Human               bool
    Pipeline            string
    Pretty              bool
    Refresh             string
    Routing             string
    Source              []string
    SourceExcludes      []string
    SourceIncludes      []string
    Timeout             time.Duration
    WaitForActiveShards string
}

func (f *DefaultBulkIndexerFlusher) Write(data []byte) (int, error) {
    return f.buf.Write(data)
}

func (f *DefaultBulkIndexerFlusher) ReadFrom(r io.Reader) (int64, error) {
    // Implement io.ReaderFrom to more efficient io.Copy, when buffering item bodies.
    return f.buf.ReadFrom(r)
}

func (f *DefaultBulkIndexerFlusher) Flush(ctx context.Context) (*esapi.Bulk, error) {
    return f.Client.Bulk(f.buf, f.Client.Bulk.WithIndex(f.Index), f.Client.WithContext(ctx))
}
```

With this approach you could wrap the flusher for instrumentation purposes. The flusher would be invoked from workers. Workers would take on the responsibility of decoding the response, and dispatching the callbacks.

Reply:
We have discussed this topic with @axw earlier today, to clarify some questions I had about the change. I've been thinking about it more during the day, and will summarize my thoughts.

One thing I don't like about this change is that it effectively moves the Bulk API parameters from the main configuration, `esutil.BulkIndexerConfig`, into the `Flusher`. For anybody who would like to set eg. an index name — which I assume is most of the users of the component —, or a pipeline — again, a frequent use case —, this would mean dealing with those in the custom component.

Another thing which bothers me is that I don't feel like restructuring the codebase significantly, ie. moving the buffer operations inside the `Flusher` (as I understand the proposal suggests). As I've said in earlier comments, I don't feel like delivering _everything and anything_ at once is a good development strategy. I would rather focus on the public API, so we don't have to break it in three weeks when we discover we need to eg. add a parameter to a method, and postponed _expanding_ the API later (which wouldn't break backwards compatibility).

I took a stab at extracting the basic operation needed for instrumentation, ie. the Bulk API call itself, into a separate component, and came up with the implementation in https://github.com/elastic/go-elasticsearch/commit/6a73580c2184f14386cf558517f2829116de3268. To me, this seems like solving the problem of wrapping this operation for instrumentation purposes. Can you please check that, @axw?

I've added an Elastic APM instrumentation into the Kafka demo in https://github.com/elastic/go-elasticsearch/commit/4ee74bbd1616faecc5e404c393a5bc1b011307f2, to verify how it behaves in reality. I've observed transactions being correctly recorded. I have not been successful with trying to capture errors: I've simulated many different types of errors — eg. sending an empty body in the bulk payload, which triggers and Elasticsearch error —, but the _Errors_ tab in the APM UI is empty. Maybe I'm doing something wrong – would you be able to check that as well, @axw?

Reply:
> One thing I don't like about this change is that it effectively moves the Bulk API parameters from the main configuration, esutil.BulkIndexerConfig, into the Flusher. For anybody who would like to set eg. an index name — which I assume is most of the users of the component —, or a pipeline — again, a frequent use case —, this would mean dealing with those in the custom component.

That change was based on the Flusher creating the request. With your approach of having the worker create the request and hand it off to the Flusher, it's fine to keep those params in the BulkIndexerConfig. With one exception: the client.

In your example: https://github.com/elastic/go-elasticsearch/commit/4ee74bbd1616faecc5e404c393a5bc1b011307f2#diff-7c873b1d59a93f95bb9ecbcc9a91c8baR103

You're passing in an elasticsearch.Client, but also passing it into the custom Flusher. The first one is not used at all in that case, which I find confusing. It would be simpler if you said: if BulkIndexerConfig.Flusher is nil, then a BulkIndexerDefaultFlusher will be constructed with a elasticsearch.NewDefaultClient. If you need to customise it, set Flusher. Then it becomes:

```go
esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
    Flusher: &BulkIndexerDefaultFlusher{Client: client},
})
```

> Another thing which bothers me is that I don't feel like restructuring the codebase significantly, ie. moving the buffer operations inside the Flusher (as I understand the proposal suggests). As I've said in earlier comments, I don't feel like delivering everything and anything at once is a good development strategy. I would rather focus on the public API, so we don't have to break it in three weeks when we discover we need to eg. add a parameter to a method, and postponed expanding the API later (which wouldn't break backwards compatibility).

This directly affects the API: where the buffer sits changes the roles and responsibilities. If it's in the worker, then you don't need the io.Writer interface on the Flusher. If it's in the Flusher, you do. I think keeping it in the worker is fine though.

> I've added an Elastic APM instrumentation into the Kafka demo in 4ee74bb, to verify how it behaves in reality. I've observed transactions being correctly recorded. I have not been successful with trying to capture errors: I've simulated many different types of errors — eg. sending an empty body in the bulk payload, which triggers and Elasticsearch error —, but the Errors tab in the APM UI is empty. Maybe I'm doing something wrong – would you be able to check that as well, @axw?

Because you're using `apm.CaptureError(context, error)`, which only works if the supplied context contains a transaction or span. If you want to send errors without any transaction/span, then you'll need to use `apm.DefaultTracer.NewError(err).Send()` instead.

This is what I was getting at before (offline), about the OnError function being separated from the calling context. If you were capturing the error in the InstrumentedFlusher, then you would use the context returned by apm.ContextWithTransaction, and it would work nicely.

Reply:
We have collaborated on this with @axw earlier today, and ultimately solved the issue in a different way; the `Flusher` interface added yestarday has been removed from the codebase.

Two new callbacks have been added to the configuration in https://github.com/elastic/go-elasticsearch/commit/aa612f9daef5f04549ab8eca251cdf20eeb755f7:

```
OnFlushStart func(context.Context) context.Context
OnFlushEnd   func(context.Context)
```

This allows to manipulate the context before and after the flush. See eg the unit test:
https://github.com/elastic/go-elasticsearch/blob/0212193aad5ee67a3879b1daa24168c705d39fbd/esutil/bulk_indexer_internal_test.go#L333-L343

An example of using them for APM instrumentation is available in:

https://github.com/elastic/go-elasticsearch/blob/32c0278b8105373cee0aaac0310a41c53306f5f3/_examples/bulk/kafka/kafka.go#L106-L112